server/src/main/java/org/opensearch/wlm/tracker/QueryGroupResourceUsage.java

```diff
@@ -0,0 +1,128 @@
+     * @param queryGroup instance
+     * @param settings {@link WorkloadManagementSettings} instance
+     * @return the overshooting limit for the resource
+     */
+    public double getReduceByFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+        return getCurrentUsage() - getNormalisedThresholdFor(queryGroup, settings);
+    }
+
+    /**
+     * initialises the member variable currentUsage
+     * @param tasks list of tasks in the query group
+     * @param timeSupplier nano time supplier
+     */
+    public void initialise(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier) {
+        this.setCurrentUsage(this.calculateResourceUsage(tasks, timeSupplier));
+    }
+
+    /**
+     * normalises configured value with respect to node level cancellation thresholds
+     * @param queryGroup instance
+     * @param settings {@link WorkloadManagementSettings} instance
+     * @return normalised value with respect to node level cancellation thresholds
+     */
+    public abstract double getNormalisedThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings);
+
+    /**
+     * calculates the current resource usage for the query group
+     * @param tasks list of tasks in the query group
+     * @param timeSupplier nano time supplier
+     */
+    protected abstract double calculateResourceUsage(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier);
+
+    /**
+     * class to store cpu usage for the query group
+     */
+    public static class QueryGroupCpuUsage extends QueryGroupResourceUsage {
+        @Override
+        public double getNormalisedThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+            return settings.getNodeLevelCpuCancellationThreshold() * queryGroup.getResourceLimits().get(ResourceType.CPU);
+        }
+
+        @Override
+        protected double calculateResourceUsage(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier) {
+            double usage = tasks.stream().mapToDouble(task -> {
+                return TaskResourceUsageCalculator.from(ResourceType.CPU).calculateFor(task, timeSupplier);
+            }).sum();
+
+            usage /= PROCESSOR_COUNT;
+            return usage;
+        }
+    }
+
+    /**
+     * class to store memory usage for the query group
+     */
+    public static class QueryGroupMemoryUsage extends QueryGroupResourceUsage {
+        @Override
+        public double getNormalisedThresholdFor(QueryGroup queryGroup, WorkloadManagementSettings settings) {
+            return settings.getNodeLevelMemoryCancellationThreshold() * queryGroup.getResourceLimits().get(ResourceType.MEMORY);
+        }
+
+        @Override
+        protected double calculateResourceUsage(List<QueryGroupTask> tasks, Supplier<Long> timeSupplier) {
+            double usage = tasks.stream().mapToDouble(task -> {
+                return TaskResourceUsageCalculator.from(ResourceType.MEMORY).calculateFor(task, timeSupplier);
```

Comment:
I am wondering if this can be combined with `TaskResourceUsageCalculator` and we can have single interface called `ResourceUsageCalculator`?

Reply:
I believe keeping this as part of object state is only going incur unnecessary overhead since these values are only needed for the calculation but will remain throughout the run of the logic (QueryGroupService). Because this reasoning will lead to having all method parameters as object state.

Reply:
Merged them into a single entity.